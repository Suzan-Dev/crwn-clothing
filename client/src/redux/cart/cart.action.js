import CartActionType from './cart.type';

export const ToggleHiddenCart = () => ({
  type: CartActionType.TOGGLE_HIDDEN_CART,
});

export const addItem = (item) => ({
  type: CartActionType.ADD_ITEM,
  payload: item,
});

export const removeItem = (item) => ({
  type: CartActionType.REMOVE_ITEM,
  payload: item,
});

export const removeItemFromCart = (item) => ({
  type: CartActionType.REMOVE_ITEM_FROM_CART,
  payload: item,
});

export const clearCart = () => ({
  type: CartActionType.CLEAR_CART,
});
