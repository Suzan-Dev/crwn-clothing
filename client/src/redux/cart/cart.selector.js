import { createSelector } from 'reselect';

const selectItem = (state) => state.cart;

export const selectCartItem = createSelector(
  [selectItem],
  (cart) => cart.cartItem
);

export const selectHidden = createSelector([selectItem], (cart) => cart.hidden);

export const selectCartItemCount = createSelector(
  [selectCartItem],
  (cartItem) => cartItem.reduce((acc, cartItem) => acc + cartItem.quantity, 0)
);

export const selectCartTotal = createSelector([selectCartItem], (cartItem) =>
  cartItem.reduce(
    (acc, cartItem) => acc + cartItem.quantity * cartItem.price,
    0
  )
);
