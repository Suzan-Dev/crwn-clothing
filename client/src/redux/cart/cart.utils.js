export const addItemToCart = (cartItem, itemToAdd) => {
  const existingItem = cartItem.find((item) => item.id === itemToAdd.id);

  if (existingItem) {
    return cartItem.map((item) =>
      item.id === itemToAdd.id ? { ...item, quantity: item.quantity + 1 } : item
    );
  } else {
    return [...cartItem, { ...itemToAdd, quantity: 1 }];
  }
};

export const removeItemFromCart = (cartItem, itemToRemove) => {
  const existingItem = cartItem.find((item) => item.id === itemToRemove.id);

  if (existingItem.quantity === 1) {
    return cartItem.filter((item) => item.id !== itemToRemove.id);
  } else {
    return cartItem.map((item) =>
      item.id === itemToRemove.id
        ? { ...item, quantity: item.quantity - 1 }
        : item
    );
  }
};
