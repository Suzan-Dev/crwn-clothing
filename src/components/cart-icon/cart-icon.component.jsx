import React from 'react';

import './cart-icon.style.scss';

import { connect } from 'react-redux';

import { ReactComponent as ShoppingIcon } from '../../assets/shopping-bag.svg';
import ToggleHiddenCart from '../../redux/cart/cart.action';

const CartIcon = ({ toggleHiddenCart }) => (
  <div className='cart-icon' onClick={toggleHiddenCart}>
    <ShoppingIcon className='shopping-icon' />
    <span className='item-count'>0</span>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  toggleHiddenCart: () => dispatch(ToggleHiddenCart()),
});

export default connect(null, mapDispatchToProps)(CartIcon);
