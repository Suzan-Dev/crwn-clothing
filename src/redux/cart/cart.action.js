import CartActionType from './cart.type';

export const ToggleHiddenCart = () => ({
  type: CartActionType.TOGGLE_HIDDEN_CART,
});

export const addItem = (item) => ({
  type: CartActionType.ADD_ITEM,
  payload: item,
});
