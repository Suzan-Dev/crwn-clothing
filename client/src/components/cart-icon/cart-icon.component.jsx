import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import './cart-icon.style.scss';

import { ReactComponent as ShoppingIcon } from '../../assets/shopping-bag.svg';
import { ToggleHiddenCart } from '../../redux/cart/cart.action';
import { selectCartItemCount } from '../../redux/cart/cart.selector';

export const CartIcon = ({ itemCount, dispatch }) => (
  <div className='cart-icon' onClick={() => dispatch(ToggleHiddenCart())}>
    <ShoppingIcon className='shopping-icon' />
    <span className='item-count'>{itemCount}</span>
  </div>
);

const mapStateToProps = createStructuredSelector({
  itemCount: selectCartItemCount,
});

export default connect(mapStateToProps)(CartIcon);
