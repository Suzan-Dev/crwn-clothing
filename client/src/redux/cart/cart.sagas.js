import { all, put, call, takeLatest } from 'redux-saga/effects';
import UserActionType from '../user/user.type';
import { clearCart } from './cart.action';

// saga worker
export function* clearCartOnSuccess() {
  yield put(clearCart());
}

// saga watcher
export function* onSignOutSuccess() {
  yield takeLatest(UserActionType.SIGN_OUT_SUCCESS, clearCartOnSuccess);
}

export function* cartSagas() {
  yield all([call(onSignOutSuccess)]);
}
