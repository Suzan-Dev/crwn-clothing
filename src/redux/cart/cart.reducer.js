import CartActionType from './cart.type';
import { addItemToCart } from './cart.utils';

const INITIAL_STATE = {
  hidden: true,
  cartItem: [],
};

const CartReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case CartActionType.TOGGLE_HIDDEN_CART:
      return {
        ...state,
        hidden: !state.hidden,
      };
    case CartActionType.ADD_ITEM:
      return {
        ...state,
        cartItem: addItemToCart(state.cartItem, action.payload),
      };
    default:
      return state;
  }
};

export default CartReducer;
