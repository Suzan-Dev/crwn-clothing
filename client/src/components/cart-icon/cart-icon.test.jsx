import React from 'react';
import { shallow } from '../../setupTests';
import { CartIcon } from './cart-icon.component';
import { ToggleHiddenCart } from '../../redux/cart/cart.action';

describe('Testing: CartIcon component!', () => {
  const mockItemCount = 1;
  const mockDispatch = jest.fn();

  const mockProps = {
    itemCount: mockItemCount,
    dispatch: mockDispatch,
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<CartIcon {...mockProps} />);
  });

  test('should render CartIcon component.', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  test('should dispatch ToggleHiddenCart action when clicked on icon.', () => {
    wrapper.find('.cart-icon').simulate('click');
    expect(mockDispatch).toHaveBeenCalledWith(ToggleHiddenCart());
  });

  test('should render itemCount properly.', () => {
    expect(wrapper.find('.item-count').text()).toBe('1');
  });
});
