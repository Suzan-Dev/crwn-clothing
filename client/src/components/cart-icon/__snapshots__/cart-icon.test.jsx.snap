// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing: CartIcon component! should render CartIcon component. 1`] = `
"<div className=\\"cart-icon\\" onClick={[Function: onClick]}>
  <ForwardRef(SvgShoppingBag) className=\\"shopping-icon\\" />
  <span className=\\"item-count\\">
    1
  </span>
</div>"
`;
